src/children/XHPChildDeclarationConsistencyTrait.hack

```diff
@@ -0,0 +1,50 @@
+/*
+ *  Copyright (c) 2004-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+use namespace \Facebook\XHP\ChildValidation as XHPChild;
+
+/** Verify that a new child declaration matches the legacy codegen. */
+trait XHPChildDeclarationConsistencyTrait {
+  require extends :x:element;
+
+  abstract protected function getChildrenDeclaration(): XHPChild\Constraint;
+
+  final private static function normalize(mixed $x): mixed {
+    // (exactly one, expr, $x) is equivalent to $x; HackC
+    // always generates this form for the top-level constraint
+    if (
+      $x is (int, int, mixed) &&
+      $x[0] === XHPChild\LegacyExpressionType::EXACTLY_ONE &&
+      $x[1] === XHPChild\LegacyConstraintType::EXPRESSION
+    ) {
+      return self::normalize($x[2]);
+    }
+
+    if ($x is (int, mixed, mixed)) {
+      return tuple($x[0], self::normalize($x[1]), self::normalize($x[2]));
+    }
+
+    return $x;
+  }
+
+  final public function validateChildren(): void {
+    $old = self::normalize($this->__xhpChildrenDeclaration());
```

Comment:
This seems like it's going to make it difficult to make the necessary runtime changes. You'll have to release a version of xhp-lib that removes knowledge of the magic runtime methods before the version of HHVM that removes this method entirely.

Reply:
Roughly, my plan is:
1. this diff
2. require this trait on every XHP class with children declarations
3. flip the source of truth to be the new declaration, but keep the assertion
4. delete the assertion, delete the syntax

Reply:
There's also a possiblity of adding a second trait, defining the same abstract method:that would also add a marker interface saying "never call __xhpChildrenDeclaration, always use this"

THe docs for that would be:
- use that new trait for all new elements
- add the migration assertion for all old elements and run tests. Maybe make something similar that does logging instead of throwing depending on your needs
- when you have no invariants/logs, delete the children declarations and switch the trait to the one with the marker interface saying to ignore __xhpChildrenDeclaration

Reply:
This is actually a bit simpler: `__xhpChildrenDeclaration()` is only emitted if there is a declaration - otherwise, the implementation comes from :x:composable-element 

hackc generated an override even if it's `any`, so we can deterministicly detect if emitted